examples/allocation-endpoint/server/main.go

```diff
@@ -0,0 +1,300 @@
+		log.Fatalf("Error building JWT access token source: %v", err)
+	}
+}
+
+func extractClusterInfo(clustersInfoStr string) ([]*ClusterInfo, error) {
+	clustersInfo := []*ClusterInfo{}
+
+	if err := json.Unmarshal([]byte(clustersInfoStr), &clustersInfo); err != nil {
+		return nil, err
+	}
+	allocationRate(clustersInfo)
+	return clustersInfo, nil
+}
+
+// Calculate the allocation rate on clusters
+func allocationRate(clustersInfo []*ClusterInfo) {
+	numEnabledClusters := 0
+	numRateControlledClusters := 0
+	totalWeight := 0
+	for _, clusterInfo := range clustersInfo {
+		if clusterInfo.AllocationWeight != minAllocationWeight {
+			numEnabledClusters++
+			totalWeight += clusterInfo.AllocationWeight
+			if clusterInfo.AllocationWeight != maxAllocationWeight {
+				numRateControlledClusters++
+			}
+		}
+	}
+	// All clusters are rate controlled
+	if numEnabledClusters == numRateControlledClusters {
+		for _, clusterInfo := range clustersInfo {
+			clusterInfo.AllocationRate = float64(clusterInfo.AllocationWeight) / float64(totalWeight)
+		}
+		return
+	}
+	// Some clusters are not rate controlled thus need to take more traffic from the rate controlled ones
+	averageRate := float64(1) / float64(numEnabledClusters)
+	existingRate := float64(0)
+	for _, clusterInfo := range clustersInfo {
+		if clusterInfo.AllocationWeight != 100 {
+			clusterInfo.AllocationRate = averageRate * float64(clusterInfo.AllocationWeight) / float64(100)
+			existingRate += clusterInfo.AllocationRate
+		}
+	}
+	for _, clusterInfo := range clustersInfo {
+		if clusterInfo.AllocationWeight == 100 {
+			clusterInfo.AllocationRate = (float64(1) - existingRate) / float64(numEnabledClusters-numRateControlledClusters)
+		}
+	}
+}
+
+func getConnection(ctx context.Context, clusterInfo *ClusterInfo) (*grpc.ClientConn, int, error) {
+	mClustersConnections.RLock()
+	clusterConnection, ok := clustersConnections[clusterInfo.Name]
+	mClustersConnections.RUnlock()
+	if !ok {
+		clusterConnection = NewClusterConns(numOfConnections)
+		mClustersConnections.Lock()
+		clustersConnections[clusterInfo.Name] = clusterConnection
+		mClustersConnections.Unlock()
+	}
+	perm := rand.Perm(numOfConnections)
+	for _, randIndex := range perm {
+		currentConn, createdTime, _ := clusterConnection.Get(randIndex)
+		if currentConn == nil || time.Since(createdTime) > connectionPoolTimeout {
+			conn, err := connectToAgonesCluster(ctx, clusterInfo)
+			if err != nil {
+				log.Error(err)
+				continue
+			}
+			clusterConnection.Set(randIndex, conn, time.Now())
+			return conn, randIndex, nil
+		}
+		return currentConn, randIndex, nil
+	}
+
+	return nil, -1, errors.Errorf("could not connect to %s with endpoint %s", clusterInfo.Name, clusterInfo.Endpoint)
+}
+
+func (s *allocationEndpointService) Allocate(ctx context.Context, req *pb.AllocationRequest) (*pb.AllocationResponse, error) {
+	allocatedClusters := make(map[string]bool)
+	var err error
```

Comment:
Why are these defined here instead of just doing `resp, err :=` on line 235?

Reply:
To return the last error, if allocation fails on all clusters.